src/jit/gentree.h

```diff
@@ -4626,24 +4636,58 @@ struct GenTreeAllocObj final : public GenTreeUnOp
 // be defined already.
 //------------------------------------------------------------------------
 
-inline bool GenTree::OperIsBlkOp() const
+inline bool GenTree::OperIsBlkOp()
 {
-    return (gtOper == GT_INITBLK || gtOper == GT_COPYBLK || gtOper == GT_COPYOBJ);
+    return (((gtOper == GT_ASG) && varTypeIsStruct(gtOp.gtOp1))
+#ifndef LEGACY_BACKEND
+            || (OperIsBlk() && (AsBlk()->Data() != nullptr))
+#endif
+                );
 }
 
 inline bool GenTree::OperIsDynBlkOp()
 {
-    return (OperIsBlkOp() && !gtGetOp2()->IsCnsIntOrI());
+    if (gtOper == GT_ASG)
+    {
+        return gtGetOp1()->OperGet() == GT_DYN_BLK;
+    }
+#ifndef LEGACY_BACKEND
+    else if (gtOper == GT_STORE_DYN_BLK)
+    {
+        return true;
+    }
+#endif
+    return false;
 }
 
-inline bool GenTree::OperIsCopyBlkOp() const
+inline bool GenTree::OperIsCopyBlkOp()
 {
-    return (gtOper == GT_COPYOBJ || gtOper == GT_COPYBLK);
+    if (gtOper == GT_ASG)
+    {
+        return (varTypeIsStruct(gtGetOp1()) && ((gtFlags & GTF_BLK_INIT) == 0));
+    }
+#ifndef LEGACY_BACKEND
+    else if (OperIsBlk())
+    {
+        return ((AsBlk()->Data() != nullptr) && ((gtFlags & GTF_BLK_INIT) == 0));
+    }
+#endif
+    return false;
 }
 
-inline bool GenTree::OperIsInitBlkOp() const
+inline bool GenTree::OperIsInitBlkOp()
 {
-    return (gtOper == GT_INITBLK);
+    if (gtOper == GT_ASG)
+    {
+        return (varTypeIsStruct(gtGetOp1()) && ((gtFlags & GTF_BLK_INIT) != 0));
+    }
+#ifndef LEGACY_BACKEND
+    else if (OperIsBlk())
+    {
+        return ((AsBlk()->Data() != nullptr) && ((gtFlags & GTF_BLK_INIT) != 0));
```

Comment:
No it is not.  What I am suggesting is to modify else if check as follows:
Note OperIsStoreBlkOp() return true when oper is GT_STORE_OBJ/BLK/DYN_BLK.

else if (OperIsStoreBlkOp())
{
   return (gtFlags & GTF_BLK_INIT) != 0;
}

Here OperIsBlk() is unnecessarily checking against GT_OBJ/BLK/DYN_BLK and later excluding them by Data() != nullptr, which is not required if we used the above.

---
In reply to: [77387651](https://github.com/dotnet/coreclr/pull/7027#discussion_r77387651) [](ancestors = 77387651,77381269)